File: memex/internal/memex/enricher.go
package memex

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"sort"
	"strings"
)

// LLMClient is the interface for LLM-based enrichment.
type LLMClient interface {
	Analyze(note Note, candidates []Note) (*EnrichmentResult, error)
}

// Enricher processes notes in the background, finding relations via LLM.
type Enricher struct {
	queue  chan string
	store  *Store
	client LLMClient
}

// NewEnricher creates an Enricher. If client is nil, enrichment is disabled.
func NewEnricher(store *Store, client LLMClient) *Enricher {
	e := &Enricher{
		queue:  make(chan string, 100),
		store:  store,
		client: client,
	}
	if client != nil {
		go e.processLoop()
	}
	return e
}

// Enqueue adds a note ID to the enrichment queue.
func (e *Enricher) Enqueue(id string) {
	if e.client == nil {
		return
	}
	select {
	case e.queue <- id:
	default:
		log.Printf("enricher: queue full, dropping note %s", id)
	}
}

func (e *Enricher) processLoop() {
	for id := range e.queue {
		if err := e.processNote(id); err != nil {
			log.Printf("enricher: failed to process note %s: %v", id, err)
		}
	}
}

func (e *Enricher) processNote(id string) error {
	note, err := e.store.Get(id)
	if err != nil {
		return err
	}

	candidates := e.findCandidates(*note, 20)
	if len(candidates) == 0 {
		return nil
	}

	result, err := e.client.Analyze(*note, candidates)
	if err != nil {
		return fmt.Errorf("LLM analysis failed: %w", err)
	}

	if len(result.Relations) > 0 {
		if err := e.store.AddRelations(id, result.Relations); err != nil {
			return fmt.Errorf("failed to add relations: %w", err)
		}
	}

	for _, sid := range result.Superseded {
		if err := e.store.UpdateStatus(sid, "superseded"); err != nil {
			log.Printf("enricher: failed to mark %s as superseded: %v", sid, err)
		}
	}

	return nil
}

// findCandidates finds related notes using tag/source overlap and optional embedding similarity.
func (e *Enricher) findCandidates(note Note, k int) []Note {
	scores := make(map[string]float64)

	// Tag overlap scoring
	tagsIdx := e.store.TagsIndex()
	for _, tag := range note.Tags {
		for _, id := range tagsIdx[tag] {
			if id != note.ID {
				scores[id] += 2.0
			}
		}
	}

	// Source overlap scoring
	sourcesIdx := e.store.SourcesIndex()
	for _, src := range note.Sources {
		key := src.Key()
		for skey, ids := range sourcesIdx {
			if strings.HasPrefix(skey, key) || strings.HasPrefix(key, skey) {
				for _, id := range ids {
					if id != note.ID {
						scores[id] += 3.0
					}
				}
			}
		}
	}

	// Embedding similarity (if available)
	noteEmb := e.store.GetEmbedding(note.ID)
	if noteEmb != nil {
		allEmbs := e.store.AllEmbeddings()
		for id, emb := range allEmbs {
			if id != note.ID {
				sim := CosineSimilarity(noteEmb, emb)
				if sim > 0.3 {
					scores[id] += sim * 5.0
				}
			}
		}
	}

	// Keyword overlap as fallback (if no other signals)
	if len(scores) == 0 {
		noteTokens := uniqueTokens(note.Content)
		summaries, err := e.store.List()
		if err == nil {
			for _, s := range summaries {
				if s.ID == note.ID {
					continue
				}
				n, err := e.store.Get(s.ID)
				if err != nil {
					continue
				}
				candTokens := uniqueTokens(n.Content)
				overlap := 0
				for t := range noteTokens {
					if candTokens[t] {
						overlap++
					}
				}
				if overlap > 2 {
					scores[s.ID] = float64(overlap)
				}
			}
		}
	}

	// Sort by score, take top-K
	type scored struct {
		id    string
		score float64
	}
	var ranked []scored
	for id, score := range scores {
		ranked = append(ranked, scored{id, score})
	}
	sort.Slice(ranked, func(i, j int) bool {
		return ranked[i].score > ranked[j].score
	})

	limit := k
	if len(ranked) < limit {
		limit = len(ranked)
	}

	var candidates []Note
	for _, r := range ranked[:limit] {
		n, err := e.store.Get(r.id)
		if err != nil {
			continue
		}
		candidates = append(candidates, *n)
	}

	return candidates
}

// --- Anthropic API Client ---

// AnthropicClient makes direct HTTP calls to the Anthropic Messages API.
type AnthropicClient struct {
	APIKey  string
	Model   string
	BaseURL string
}

// NewAnthropicClient creates a client with defaults.
func NewAnthropicClient(apiKey, model string) *AnthropicClient {
	if model == "" {
		model = "claude-haiku-4-5-20251001"
	}
	return &AnthropicClient{
		APIKey:  apiKey,
		Model:   model,
		BaseURL: "https://api.anthropic.com/v1/messages",
	}
}

type anthropicRequest struct {
	Model     string             `json:"model"`
	MaxTokens int                `json:"max_tokens"`
	Messages  []anthropicMessage `json:"messages"`
}

type anthropicMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type anthropicResponse struct {
	Content []struct {
		Text string `json:"text"`
	} `json:"content"`
}

// Analyze sends a note and candidates to Claude API for relation analysis.
func (c *AnthropicClient) Analyze(note Note, candidates []Note) (*EnrichmentResult, error) {
	prompt := buildEnrichmentPrompt(note, candidates)

	reqBody := anthropicRequest{
		Model:     c.Model,
		MaxTokens: 1024,
		Messages: []anthropicMessage{
			{Role: "user", Content: prompt},
		},
	}

	body, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", c.BaseURL, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", c.APIKey)
	req.Header.Set("anthropic-version", "2023-06-01")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("API request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		respBody, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("API returned status %d: %s", resp.StatusCode, string(respBody))
	}

	var apiResp anthropicResponse
	if err := json.NewDecoder(resp.Body).Decode(&apiResp); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if len(apiResp.Content) == 0 {
		return &EnrichmentResult{}, nil
	}

	var result EnrichmentResult
	text := apiResp.Content[0].Text

	// Extract JSON from response (handle markdown code blocks)
	jsonStr := extractJSON(text)
	if err := json.Unmarshal([]byte(jsonStr), &result); err != nil {
		return nil, fmt.Errorf("failed to parse enrichment result: %w (response: %s)", err, text)
	}

	return &result, nil
}

func buildEnrichmentPrompt(note Note, candidates []Note) string {
	var sb strings.Builder

	sb.WriteString("Analyze the NEW note below and find relationships to EXISTING notes.\n\n")
	sb.WriteString("NEW NOTE:\n")
	sb.WriteString(fmt.Sprintf("ID: %s\nType: %s\nContent: %s\nTags: %s\n\n",
		note.ID, note.Type, note.Content, strings.Join(note.Tags, ", ")))

	sb.WriteString("EXISTING NOTES:\n")
	for _, c := range candidates {
		sb.WriteString(fmt.Sprintf("- ID: %s | Type: %s | Status: %s | Content: %s\n",
			c.ID, c.Type, c.Status, c.Content))
	}

	sb.WriteString(`
Respond with ONLY a JSON object (no markdown, no explanation):
{
  "relations": [
    {"target_id": "<existing_note_id>", "type": "<relation_type>"}
  ],
  "superseded": ["<note_id_that_new_note_supersedes>"]
}

Relation types: relates_to, depends_on, contradicts, supersedes, elaborates, blocks
Rules:
- Only include relations with high confidence
- "superseded" lists IDs of existing notes that the new note completely replaces
- If no relations found, return {"relations": [], "superseded": []}
- target_id in relations must reference existing note IDs only
`)

	return sb.String()
}

func extractJSON(text string) string {
	// Try to find JSON in markdown code block
	if idx := strings.Index(text, "```json"); idx >= 0 {
		text = text[idx+7:]
		if end := strings.Index(text, "```"); end >= 0 {
			return strings.TrimSpace(text[:end])
		}
	}
	if idx := strings.Index(text, "```"); idx >= 0 {
		text = text[idx+3:]
		if end := strings.Index(text, "```"); end >= 0 {
			return strings.TrimSpace(text[:end])
		}
	}
	// Try to find raw JSON
	if idx := strings.Index(text, "{"); idx >= 0 {
		depth := 0
		for i := idx; i < len(text); i++ {
			switch text[i] {
			case '{':
				depth++
			case '}':
				depth--
				if depth == 0 {
					return text[idx : i+1]
				}
			}
		}
	}
	return text
}

File: memex/src/enricher.ts
import type { Note, Relation, EnrichmentResult } from "./types.js";
import { sourceKey } from "./types.js";
import { Store } from "./store.js";
import { cosineSimilarity } from "./search.js";
import { extractJSON } from "./utils.js";

export interface LLMClient {
  analyze(note: Note, candidates: Note[]): Promise<EnrichmentResult>;
}

export class Enricher {
  private queue: string[] = [];
  private processing = false;
  private store: Store;
  private client: LLMClient | null;

  constructor(store: Store, client: LLMClient | null) {
    this.store = store;
    this.client = client;
  }

  enqueue(id: string): void {
    if (!this.client) return;
    this.queue.push(id);
    this.processNext();
  }

  private async processNext(): Promise<void> {
    if (this.processing || this.queue.length === 0) return;
    this.processing = true;

    while (this.queue.length > 0) {
      const id = this.queue.shift()!;
      try {
        await this.processNote(id);
      } catch (err) {
        console.error(`enricher: failed to process note ${id}:`, err);
      }
    }

    this.processing = false;
  }

  private async processNote(id: string): Promise<void> {
    const note = this.store.get(id);
    const candidates = this.findCandidates(note, 20);
    if (candidates.length === 0) return;

    const result = await this.client!.analyze(note, candidates);

    if (result.relations.length > 0) {
      this.store.addRelations(id, result.relations);
    }

    for (const sid of result.superseded ?? []) {
      try {
        this.store.updateStatus(sid, "superseded");
      } catch (err) {
        console.error(`enricher: failed to mark ${sid} as superseded:`, err);
      }
    }
  }

  private findCandidates(note: Note, k: number): Note[] {
    const scores = new Map<string, number>();
    const tags = this.store.tagsIndex();
    const sources = this.store.sourcesIndex();

    // Tag overlap
    for (const tag of note.tags) {
      for (const id of tags[tag] ?? []) {
        if (id !== note.id) scores.set(id, (scores.get(id) ?? 0) + 2);
      }
    }

    // Source overlap
    for (const src of note.sources) {
      const key = sourceKey(src);
      for (const [skey, ids] of Object.entries(sources)) {
        if (skey.startsWith(key) || key.startsWith(skey)) {
          for (const id of ids) {
            if (id !== note.id) scores.set(id, (scores.get(id) ?? 0) + 3);
          }
        }
      }
    }

    // Embedding similarity
    const noteEmb = this.store.getEmbedding(note.id);
    if (noteEmb) {
      for (const [id, emb] of Object.entries(this.store.allEmbeddings())) {
        if (id !== note.id) {
          const sim = cosineSimilarity(noteEmb, emb);
          if (sim > 0.3) scores.set(id, (scores.get(id) ?? 0) + sim * 5);
        }
      }
    }

    // Sort and take top-k
    const ranked = [...scores.entries()]
      .sort((a, b) => b[1] - a[1])
      .slice(0, k);

    return ranked
      .map(([id]) => { try { return this.store.get(id); } catch { return null; } })
      .filter((n): n is Note => n !== null);
  }
}

// --- Agent SDK Client ---

export class AgentSDKClient implements LLMClient {
  private authToken?: string;
  private apiKey?: string;

  constructor(authToken?: string, apiKey?: string) {
    this.authToken = authToken;
    this.apiKey = apiKey;
  }

  async analyze(note: Note, candidates: Note[]): Promise<EnrichmentResult> {
    const { query } = await import("@anthropic-ai/claude-agent-sdk");
    const prompt = buildEnrichmentPrompt(note, candidates);

    // Strip CLAUDECODE env to avoid "nested session" error when
    // the MCP server (spawned by Claude Code) calls the Agent SDK.
    const env: Record<string, string | undefined> = { ...process.env };
    delete env.CLAUDECODE;

    // Auth priority:
    // 1. process.env (CLAUDE_CODE_OAUTH_TOKEN or ANTHROPIC_API_KEY)
    // 2. config.auth_token → CLAUDE_CODE_OAUTH_TOKEN
    // 3. config.api_key → ANTHROPIC_API_KEY
    if (!env.CLAUDE_CODE_OAUTH_TOKEN && !env.ANTHROPIC_API_KEY) {
      if (this.authToken) {
        env.CLAUDE_CODE_OAUTH_TOKEN = this.authToken;
      } else if (this.apiKey) {
        env.ANTHROPIC_API_KEY = this.apiKey;
      }
    }
    if (!env.CLAUDE_CODE_OAUTH_TOKEN && !env.ANTHROPIC_API_KEY) {
      console.error("enricher: no auth found. Set via: memex config set auth_token <token> or memex config set api_key <key>");
      return { relations: [], superseded: [] };
    }

    let resultText = "";
    for await (const message of query({
      prompt,
      options: {
        allowedTools: [],
        maxTurns: 1,
        tools: [],
        model: "claude-haiku-4-5-20251001",
        effort: "low" as const,
        persistSession: false,
        env,
      },
    })) {
      const msg = message as any;
      // Collect assistant text
      if (msg.type === "assistant" && msg.message?.content) {
        for (const block of msg.message.content) {
          if (block.type === "text") resultText += block.text;
        }
      }
      // Also check for result message
      if ("result" in msg && typeof msg.result === "string") {
        resultText = msg.result;
      }
    }

    if (!resultText) return { relations: [], superseded: [] };

    const json = extractJSON(resultText);
    try {
      return JSON.parse(json);
    } catch {
      console.error("enricher: failed to parse result:", resultText);
      return { relations: [], superseded: [] };
    }
  }
}

function buildEnrichmentPrompt(note: Note, candidates: Note[]): string {
  let prompt = "Analyze the NEW note below and find relationships to EXISTING notes.\n\n";
  prompt += `NEW NOTE:\nID: ${note.id}\nType: ${note.type}\nContent: ${note.content}\nTags: ${note.tags.join(", ")}\n\n`;
  prompt += "EXISTING NOTES:\n";
  for (const c of candidates) {
    prompt += `- ID: ${c.id} | Type: ${c.type} | Status: ${c.status} | Content: ${c.content}\n`;
  }
  prompt += `
Respond with ONLY a JSON object (no markdown, no explanation):
{
  "relations": [
    {"target_id": "<existing_note_id>", "type": "<relation_type>"}
  ],
  "superseded": ["<note_id_that_new_note_supersedes>"]
}

Relation types: relates_to, depends_on, contradicts, supersedes, elaborates, blocks
Rules:
- Only include relations with high confidence
- "superseded" lists IDs of existing notes that the new note completely replaces
- If no relations found, return {"relations": [], "superseded": []}
- target_id in relations must reference existing note IDs only
`;
  return prompt;
}

